Sources/SourceKit/sourcekitd/SemanticRefactoring.swift

```diff
@@ -0,0 +1,157 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+import LanguageServerProtocol
+import TSCBasic
+import sourcekitd
+
+/// Detailed information about the result of a specific refactoring operation.
+///
+/// Wraps the information returned by sourcekitd's `semantic_refactoring` request, such as the necessary edits and placeholder locations.
+struct SemanticRefactoring {
+
+  /// The title of the refactoring action.
+  var title: String
+
+  /// The resulting `WorkspaceEdit` of a `semantic_refactoring` request.
+  var edit: WorkspaceEdit
+
+  init(_ title: String, _ edit: WorkspaceEdit) {
+    self.title = title
+    self.edit = edit
+  }
+
+  /// Create a `SemanticRefactoring` from a sourcekitd response dictionary, if possible.
+  ///
+  /// - Parameters:
+  ///   - title: The title of the refactoring action.
+  ///   - dict: Response dictionary to extract information from.
+  ///   - url: The client URL that triggered the `semantic_refactoring` request.
+  ///   - keys: The sourcekitd key set to use for looking up into `dict`.
+  init?(_ title: String, _ dict: SKResponseDictionary, _ url: URL, _ keys: sourcekitd_keys) {
+    guard let categorizedEdits: SKResponseArray = dict[keys.categorizededits] else {
+      return nil
+    }
+
+    var textEdits = [TextEdit]()
+
+    categorizedEdits.forEach { _, value in
+      guard let edits: SKResponseArray = value[keys.edits] else {
+        return false
+      }
+      edits.forEach { _, value in
+        if let startLine: Int = value[keys.line],
+           let startColumn: Int = value[keys.column],
+           let endLine: Int = value[keys.endline],
+           let endColumn: Int = value[keys.endcolumn],
+           let text: String = value[keys.text]
+        {
+          // The LSP is zero based, but semantic_refactoring is one based.
+          let startPosition = Position(line: startLine - 1, utf16index: startColumn - 1)
```

Comment:
These columns need to be converted from UTF-8 (used by sourcekitd) to UTF-16 (used by LSP).  You can use the snapshot's `positionOf(zeroBasedLine:utf8Column:)` to get a unicode-correct position.